event/consume.go

```diff
@@ -0,0 +1,112 @@
+package event
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ONSdigital/dp-import-cantabular-dataset/schema"
+	importapi "github.com/ONSdigital/dp-import-api/models"
+	"github.com/ONSdigital/dp-api-clients-go/dataset"
+
+	kafka "github.com/ONSdigital/dp-kafka/v2"
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+// Consume converts messages to event instances, and pass the event to the provided handler.
+func (p *Processor) Consume(ctx context.Context, cg kafka.IConsumerGroup, h Handler) {
+	// consume loop, to be executed by each worker
+	var consume = func(workerID int) {
+		for {
+			select {
+			case msg, ok := <-cg.Channels().Upstream:
+				if !ok {
+					log.Info(ctx, "upstream channel closed - closing event consumer loop", log.Data{"worker_id": workerID})
+					return
+				}
+
+				ctx := context.Background()
+
+				if errs := p.processMessage(ctx, msg, h); len(errs) != 0{
+					var errdata []map[string]interface{}
+
+					for _, err := range errs {
+						errdata = append(errdata, map[string]interface{}{
+							"error":    err.Error(),
+							"log_data": unwrapLogData(err),
+							"status_code": statusCode(err),
+						})
+					}
+
+					log.Event(ctx, "failed to process message", log.ERROR, log.Data{
+						"errors": errdata,
+					})
+				}
+
+				msg.Release()
+			case <-cg.Channels().Closer:
+				log.Info(ctx, "closer channel closed - closing event consumer loop ", log.Data{"worker_id": workerID})
+				return
+			}
```

Comment:
Can we also add a use case of the `<-ctx.Done()` when the parent context no longer exists.

Reply:
Sure. Sounds like something that should be included in the blueprint.